### DISCUSSION TITLE
Why doesn't npm cli have autocompletion?

### DISCUSSION THREAD

**Discussion by vdegenne** [npm/feedback#1060] [CLI]
For me it is absolutely mind blowing that npm still doesn't have autocompletion for package names in 2024.
By "package names" I don't mean searching for available packages online when I press TAB, but just currently installed packages.
For instance if I type `npm update rollup-plug` and I have `rollup-plugin-node-resolve` and `rollup-plugin-typescript` installed in my project then the terminal should propose both packages for autocompletion when I press TAB (similar to `npm run ...` suggesting defined scripts), it works for scripts but not for installed package. Why do we have to type everything? So much unnecessary typing, I just don't understand why such a popular (and quite old now) tool doesn't have such basic features.
Not only autocompletion would help us to code faster but could also serve as a small visual hint, for instance if I type `npm info jq` and press TAB, if I don't see `jquery` being autocompleted then I would know it's not installed in my project.

Does anyone know if `yarn`, `pnpm`, `bun` or another sibling of npm have this feature available? 

(ps: do not even get me started about `npm` not showing the version of a package I just installed using `npm install <package_name>` (such a joke))

**Comment by Malvineous** [comment]
Isn't this up to the shell you're using?  There are dozens of different shells and they each support autocompletion in different ways, so maybe you need to get your preferred shell to add support for npm?

For instance [this project](https://github.com/lukechilds/zsh-better-npm-completion) looks to have been around for eight years now and adds the kind of autocompletion you're looking for to zsh.

**Reply by vdegenne** [reply]
Thanks a lot, I've never heard of it before. I'll have to try that next week! It looks promising.
What I meant is why isn't that type of auto-completion automatically installed? For instance I get auto-completion for run commands so I assume they get installed when I install npm on a Linux distribution.
Why can't they get inspiration from this and natively provide a better user experience?
Anyway Thanks again for your help.

**Reply by Malvineous** [reply]
It's not automatically installed for a number of reasons.  For one, npm don't know what shell you're using.  Should they install the bash autocompletions, or csh ones, or ksh or zsh, or what?  Wikipedia has a [list of common shells](https://en.wikipedia.org/wiki/Comparison_of_command_shells) but there are dozens and which ones would you support?

Also, npm generally doesn't mange its own installation (at least under Linux) - that's handled by the distribution.  So if you're running Gentoo Linux, it's Gentoo's responsibility to install the autocompletion scripts.  If you're on Ubuntu, it's up to them.  This is because each distribution decides where to store the files that handle the autocompletion process, so it's just not practical for npm to know where each of the hundreds of different Linux distributions keep their files.  And again, this location is not just different between distributions, it's also different for each of those dozens of shells.  So zsh autocompletions might be in different paths between Linux Mint and Debian, but even on only Debian, bash and zsh will store the autocompletion scripts in different folders.

So hopefully you can see why it's just not an option for npm to offer this - there are too many unknowns.  Some projects provide autocompletion scripts for a couple of popular shells, but it still requires the distribution to actually install them, and many don't (because who wants all these csh scripts on their system when you only use bash?)

Installation of autocompletion scripts is thus either left up to the end user, only installing ones they want to use for the shells they are using, or in some cases by the shell itself - I believe zsh comes with an extensive library of autocompletion scripts (including no doubt for many programs you will never use) so if autocompletion is important to you, using zsh is probably a good start.
